File: app/src/main/java/remix/myplayer/helper/ShakeDetector.java
package remix.myplayer.helper;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.Intent;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Handler;

import remix.myplayer.App;
import remix.myplayer.service.Command;
import remix.myplayer.service.MusicService;
import remix.myplayer.util.LogUtil;

/**
 * @ClassName
 * @Description
 * @Author Xiaoborui
 * @Date 2017/1/22 10:27
 */

public class ShakeDetector extends ContextWrapper implements SensorEventListener{
    private SensorManager mSensorManager;
    private Sensor mSensor;

    //发送命令间隔
    private static final int UPDATE_INTERVAL_TIME = 100;
    //加速度阈值
    private static final int SPEED_THRESHOLD = 400;
    //检测间隔
    private static final int DETECTION_THRESHOLD = 100;

    private long mLastUpdateTime;
    private float mLastX;
    private float mLastY;
    private float mLastZ;
    private boolean mDetect = false;
    private Handler mHandler = new Handler();
    private Runnable mRunnable = () -> sendBroadcast(new Intent(MusicService.ACTION_CMD)
            .putExtra("Control", Command.NEXT));

    private static ShakeDetector mInstance;
    private ShakeDetector(Context context){
        super(context);
    }

    public synchronized static ShakeDetector getInstance(){
        if(mInstance == null){
            mInstance = new ShakeDetector(App.getContext());
        }
        return mInstance;
    }

    public void beginListen(){
        mDetect = true;
        //已经开始监听
        if(mSensor != null)
            return;
        if(mSensorManager == null)
            mSensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
        if(mSensor == null)
            mSensor = mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        mSensorManager.registerListener(this, mSensor, SensorManager.SENSOR_DELAY_UI);
    }

    public void stopListen(){
        mDetect = false;
        if(mSensorManager == null)
            mSensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);

        if(mSensor == null){
            mSensor = mSensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
        }
        mSensorManager.unregisterListener(this,mSensor);
        mSensorManager = null;
        mSensor = null;
        mHandler.removeCallbacks(mRunnable);
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if(!mDetect)
            return;
        long currentUpdateTime = System.currentTimeMillis();
        long timeInterval = currentUpdateTime - mLastUpdateTime;
        if(timeInterval < DETECTION_THRESHOLD){
            return;
        }
        mLastUpdateTime = currentUpdateTime;
        // 计算传感器差值
        float[] values = event.values;
        float x = values[0];
        float y = values[1];
        float z = values[2];
        float deltaX = x - mLastX;
        float deltaY = y - mLastY;
        float deltaZ = z - mLastZ;
        double speed = (Math.sqrt(deltaX * deltaX + deltaY * deltaY + deltaZ * deltaZ) / timeInterval) * 1000;
        LogUtil.d("ShakeDetector","speed: " + speed + " interval: " + timeInterval + " sensorType: " + event.sensor.getType());
        if(speed > SPEED_THRESHOLD){
            mHandler.removeCallbacks(mRunnable);
            mHandler.postDelayed(mRunnable,UPDATE_INTERVAL_TIME);
        }
        mLastX = x;
        mLastY = y;
        mLastZ = z;

    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {

    }
}

File: app/src/main/java/remix/myplayer/helper/ShakeDetector.kt
package remix.myplayer.helper

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Handler
import android.os.Looper
import remix.myplayer.App
import remix.myplayer.service.Command
import remix.myplayer.util.Util.sendCMDLocalBroadcast

/**
 * @ClassName
 * @Description
 * @Author Xiaoborui
 * @Date 2017/1/22 10:27
 */

class ShakeDetector private constructor() : SensorEventListener, Runnable {
  private var sensorManager: SensorManager? = null
  private var sensor: Sensor? = null

  private var lastDetectTime: Long = 0
  private var lastPostTime: Long = 0
  private var lastX: Float = 0.toFloat()
  private var lastY: Float = 0.toFloat()
  private var lastZ: Float = 0.toFloat()
  private var begin = false
  private val handler = Handler(Looper.getMainLooper())

  fun beginListen() {
    if (begin)
      return
    begin = true
    sensorManager = App.getContext().getSystemService(Context.SENSOR_SERVICE) as SensorManager
    sensor = sensorManager?.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)
    sensorManager?.registerListener(this, sensor, SensorManager.SENSOR_DELAY_UI)
  }

  fun stopListen() {
    begin = false
    sensorManager?.unregisterListener(this, sensor)
    sensorManager = null
    sensor = null
    handler.removeCallbacks(this)
  }

  override fun onSensorChanged(event: SensorEvent) {
    if (!begin)
      return
    val currentTime = System.currentTimeMillis()
    val detectInterval = currentTime - lastDetectTime
    if (detectInterval < DETECTION_THRESHOLD) {
      return
    }
    lastDetectTime = currentTime
    // 计算传感器差值
    val values = event.values
    val x = values[0]
    val y = values[1]
    val z = values[2]
    val deltaX = x - lastX
    val deltaY = y - lastY
    val deltaZ = z - lastZ
    val speed = Math.sqrt((deltaX * deltaX + deltaY * deltaY + deltaZ * deltaZ).toDouble()) / detectInterval * 1000
    if (speed > SPEED_THRESHOLD) {
      val postInterval = currentTime - lastPostTime
      if (postInterval > POST_THRESHOLD) {
        handler.removeCallbacks(this)
        handler.post(this)
        lastPostTime = currentTime
      }
    }
    lastX = x
    lastY = y
    lastZ = z

  }

  override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {

  }

  override fun run() {
    sendCMDLocalBroadcast(Command.NEXT)
  }

  companion object {
    //加速度阈值
    private const val SPEED_THRESHOLD = 300
    //检测间隔
    private const val DETECTION_THRESHOLD = 100
    //发送指令间隔
    private const val POST_THRESHOLD = 1000L

    @JvmStatic
    @Synchronized
    fun getInstance() = Holder.INSTANCE

  }

  private object Holder {
    val INSTANCE = ShakeDetector()
  }
}
